src/engine/Sort.cpp

```diff
@@ -52,30 +55,131 @@ std::string Sort::getDescriptor() const {
   Engine::sort(idTable, sortColumnIndices_);
 
-  // Don't report missed timeout check because sort is not cancellable
+  // Don't report missed timeout check because the in-memory sort is currently
+  // not cancellable.
+  //
+  // TODO: make the in-memory sort cancellable.
   cancellationHandle_->resetWatchDogState();
   checkCancellation();
 
-  AD_LOG_DEBUG << "Sort result computation done." << endl;
+  // Return the sorted result.
   return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
 }
 
+// _____________________________________________________________________________
+Result Sort::computeResultExternalSort(std::shared_ptr<const Result> subRes,
+                                       bool requestLaziness) {
+  runtimeInfo().addDetail("is-external", "true");
+
+  // Create a unique temporary filename for the external sorter in the index
+  // directory.
+  const std::string& onDiskBase =
+      getExecutionContext()->getIndex().getOnDiskBase();
+  std::ostringstream oss;
+  oss << onDiskBase << ".sort-" << reinterpret_cast<uintptr_t>(this) << "-"
+      << std::this_thread::get_id() << ".dat";
+  std::string tempFilename = oss.str();
+
+  // Get memory limit from the same runtime parameters that we use for the
+  // materialized view writer.
+  //
+  // TODO: Give this runime parameter a more general name.
+  ad_utility::MemorySize memoryLimit =
+      getRuntimeParameter<&RuntimeParameters::materializedViewWriterMemory_>();
+
+  // Get the number of columns from the subtree (this works for both lazy and
+  // materialized results).
+  size_t numColumns = subtree_->getResultWidth();
+
+  // Create the external sorter with our dynamic column comparator. We use a
+  // shared pointer so the sorter survives when the lambda below captures it.
+  using Sorter = ad_utility::CompressedExternalIdTableSorter<SortByColumns, 0>;
+  auto sorter = std::make_shared<Sorter>(
+      tempFilename, numColumns, memoryLimit, allocator(),
+      ad_utility::DEFAULT_BLOCKSIZE_EXTERNAL_ID_TABLE,
+      SortByColumns{sortColumnIndices_});
+
+  // Push all rows into the sorter, handling both lazy and materialized results.
+  // For materialized results, we copy the local vocab, for lazy results we
+  // merge the local vocabs from the individual blocks.
+  auto mergedLocalVocab = std::make_shared<LocalVocab>();
+  if (subRes->isFullyMaterialized()) {
+    sorter->pushBlock(subRes->idTable());
+    *mergedLocalVocab = subRes->getCopyOfLocalVocab();
+  } else {
+    for (Result::IdTableVocabPair& pair : subRes->idTables()) {
+      checkCancellation();
+      sorter->pushBlock(pair.idTable_);
+      mergedLocalVocab->mergeWith(pair.localVocab_);
+    }
+  }
+
+  // If laziness is not requested, materialize the result.
+  if (!requestLaziness) {
+    IdTable result{numColumns, allocator()};
+    for (auto& block : sorter->getSortedBlocks<0>()) {
+      checkCancellation();
+      result.insertAtEnd(block);
```

Comment:
We already know how many elements there are in total, so we should definitely reserve this before inserting everything (also this is rather inefficient, we should expose an interface for this)